### PR TITLE
PHPC-1050: Command cursor should not invoke getMore at execution

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-1050

### CODE REVIEW

File: tests/cursor/bug1050-002.phpt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+$cursor = $manager->executeReadCommand(DATABASE_NAME, $cmd);
+printf("Executing command took %0.6f seconds\n", microtime(true) - $start);
+
+$it = new IteratorIterator($cursor);
+
+printf("Current position is valid: %s\n", $it->valid() ? 'yes' : 'no');
+
+$start = microtime(true);
+$it->next();
+printf("Advancing cursor took %0.6f seconds\n", microtime(true) - $start);
```
Comment (jmikola) [inline tests/cursor/bug1050-002.phpt]: Unlike the first test, this one omits the initial call to `rewind()` and tests that we preserve existing behavior by not requiring the rewind handler to ever be invoked (only relevant when wrapping a cursor with IteratorIterator).

File: src/MongoDB/Cursor.c [hunk 2]
```diff
@@ -96,7 +96,15 @@ static void php_phongo_cursor_iterator_move_forward(zend_object_iterator *iter T
...
-	cursor->current++;
+
+	/* If the cursor has already advanced, increment its position. Otherwise,
+	 * the first call to mongoc_cursor_next() will be made below and we should
+	 * leave its position at zero. */
+	if (cursor->advanced) {
+		cursor->current++;
+	} else {
+		cursor->advanced = true;
+	}
```
Comment (jmikola) [inline src/MongoDB/Cursor.c]: Future improvement: refactor `phongo_cursor_advance_and_check_for_error()` so that it can take an optional `bson_t *`. Then `phongo_execute_query()` and `php_phongo_cursor_iterator_rewind()` could pass in `null`, since they only care about error checking and don't need to access the document. Meanwhile, `php_phongo_cursor_iterator_move_forward()` could pass in a `bson_t *` and, on success, use that to populate the current element, as is done with:

```c
php_phongo_bson_to_zval_ex(bson_get_data(doc), doc->len, &cursor->visitor_data);
```

File: php_phongo.c [hunk 3]
```diff
@@ -788,7 +763,8 @@ int phongo_execute_query(mongoc_client_t *client, const char *namespace, zval *z
 		mongoc_cursor_set_max_await_time_ms(cursor, query->max_await_time_ms);
 	}
 
-	if (!phongo_advance_cursor_and_check_for_error(cursor TSRMLS_CC)) {
+	if (!phongo_cursor_advance_and_check_for_error(cursor TSRMLS_CC)) {
+		mongoc_cursor_destroy(cursor);
```
Comment (derickr) [inline php_phongo.c]: It should *not* destroy the cursor, as ``phongo_cursor_advance_and_check_for_error`` already does that.

File: php_phongo.c [hunk 4]
```diff
@@ -916,15 +892,33 @@ int phongo_execute_command(mongoc_client_t *client, php_phongo_command_type_t ty
...
-	if (!phongo_advance_cursor_and_check_for_error(cmd_cursor TSRMLS_CC)) {
-		/* If an error is found, the cmd_cursor is destroyed already */
-		return false;
-	}
-
 	phongo_cursor_init_for_command(return_value, client, cmd_cursor, db, zcommand, zreadPreference TSRMLS_CC);
 	return true;
 } /* }}} */
 
+/* Advance the cursor and return whether there is an error. On error, false is
```
Comment (derickr) [inline php_phongo.c]: Moving this down here made it impossible to see the changes in the diff, can you please put it back in its original location? That way, I would have spotted that you removed the ``mongoc_cursor_destroy(cursor);`` from it.

File: src/MongoDB/BulkWrite.c [hunk 5]
```diff
@@ -188,7 +188,7 @@ static PHP_METHOD(BulkWrite, __construct)
 		ordered = php_array_fetchc_bool(options, "ordered");
 	}
 
-	intern->bulk = phongo_bulkwrite_init(ordered);
+	intern->bulk = mongoc_bulk_operation_new(ordered);
```
Comment (derickr) [inline src/MongoDB/BulkWrite.c]: This is really unrelated to the patch, you just removed ``phongo_bulkwrite_init`` as it only did this one thing. I would suggest you don't do this in this single commit.

File: src/MongoDB/Cursor.c [hunk 6]
```diff
@@ -117,6 +125,16 @@ static void php_phongo_cursor_iterator_rewind(zend_object_iterator *iter TSRMLS_
 	php_phongo_cursor_t        *cursor = cursor_it->cursor;
 	const bson_t               *doc;
 
+	/* If the cursor was never advanced (e.g. command cursor), do so now */
+	if (!cursor->advanced) {
+		cursor->advanced = true;
+
+		if (!phongo_cursor_advance_and_check_for_error(cursor->cursor TSRMLS_CC)) {
+			/* Exception should already have been thrown */
```
Comment (derickr) [inline src/MongoDB/Cursor.c]: Yes, but shouldn't you destroy the cursor now?

File: php_phongo.c (hunk 3 above)
Comment (jmikola) [reply php_phongo.c]: `phongo_cursor_advance_and_check_for_error()` no longer destroys the cursor. That change was made because we don't want to destroy it from within the iterator handlers.

File: src/MongoDB/Cursor.c (hunk 6 above)
Comment (jmikola) [reply src/MongoDB/Cursor.c]: Once we create a `MongoDB\Driver\Cursor` with the `mongoc_cursor_t`, the only place we destroy the libmongoc cursor is from `php_phongo_cursor_free_object()`.

**Comment by jmikola** [comment]
The bulk write refactoring was split out into #731.
